ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java

```diff
@@ -13,119 +13,538 @@
-    // Empty blob ID
+    blobIdLikes.add(buildBlobIdLike((short) (BLOB_ID_V2 + 1), referenceFlag, referenceDatacenterId, referenceAccountId,
+        referenceContainerId, referencePartitionId, goodUUID.length(), goodUUID));
+    // Empty blobId
     blobIdLikes.add("");
     // short Blob ID
     blobIdLikes.add("AA");
 
     for (String blobIdLike : blobIdLikes) {
       try {
-        new BlobId(blobIdLike, clusterMap);
-        fail("Expected blob ID creation to fail with blob ID string " + blobIdLike);
+        BlobId.fromStringId(blobIdLike, referenceClusterMap);
+        fail("Expected blobId creation to fail with blobId string " + blobIdLike);
       } catch (Exception e) {
+        // expected
       }
     }
   }
 
   /**
-   * Build a string that resembles a blob ID, but with possibly invalid sections.
-   * @param version the version number to use in the blob ID
-   * @param partitionId the partition ID to use in the blob ID
-   * @param uuidLength the UUID length to use in the blob ID
-   * @param uuidLike the UUID to use in the blob ID
-   * @return a base-64 encoded {@link String} representing the blob ID
+   * Build a string that resembles a blobId, but with certain fields possibly set to legacy values.
+   * @param version The version number to be embedded in the blobId.
+   * @param flag The flag to be embedded in the blobId.
+   * @param datacenterId The datacenter id to be embedded in the blobId.
+   * @param accountId The account id to be embedded in the blobId.
+   * @param containerId The container id to be embedded in the blobId.
+   * @param partitionId The partition id to be embedded in the blobId.
+   * @param uuidLength The length of the uuid.
+   * @param uuidLike The UUID to be embedded in the blobId.
+   * @return a base-64 encoded {@link String} representing the blobId.
    */
-  private String buildBlobIdLike(short version, PartitionId partitionId, int uuidLength, String uuidLike) {
-    final int idLength = 2 + partitionId.getBytes().length + 4 + uuidLike.length();
-    ByteBuffer idBuf = ByteBuffer.allocate(idLength);
-    idBuf.putShort(version);
+  private String buildBlobIdLike(short version, Byte flag, Byte datacenterId, Short accountId, Short containerId,
+      PartitionId partitionId, int uuidLength, String uuidLike) {
+    int idLength;
+    ByteBuffer idBuf;
+    switch (version) {
+      case BLOB_ID_V1:
+        idLength = 2 + partitionId.getBytes().length + 4 + uuidLike.length();
+        idBuf = ByteBuffer.allocate(idLength);
+        idBuf.putShort(version);
+        break;
+
+      case BLOB_ID_V2:
+        idLength = 2 + 1 + 1 + 2 + 2 + partitionId.getBytes().length + 4 + uuidLike.length();
+        idBuf = ByteBuffer.allocate(idLength);
+        idBuf.putShort(version);
+        idBuf.put(flag);
+        idBuf.put(datacenterId);
+        idBuf.putShort(accountId);
+        idBuf.putShort(containerId);
+        break;
+
+      default:
```

Comment:
This will build some blobIdLike with a bad version. The bad-version id will fail a test later.